### PR TITLE
feature: Create helm charts for kubernetes deployment. Touches #4504

### PR DESCRIPTION
**Pull request by kbetanski**
```
feature: Create helm charts for kubernetes deployment

Creates two helm charts for on-premises. One is for service only and the second
for fast deployment of the whole stack required to run cloud services. Every
chart has its own README.md file with information about installation and how to
verify the deployment with tests.

Stack chart uses the service's chart as a dependency as well as bitnami/mysql
and bitnami/redis charts. It's insecure and not persistent by default, there is
a warning with that information added to the readme file. It can be used for
local development in minikube as well, the only requirement is having access to
CKEditor's container registry and license key of on-premises.

Touches: cksource/cs#4504.
```
Touches: cksource/cs#4504

### CODE REVIEW

File: kubernetes/helm/ckeditor-cs-stack/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+REGISTRY_USERNAME=
+# Your container registry authentication token
+REGISTRY_TOKEN=
+
+kubectl create secret docker-registry docker-cke-cs-com --docker-username $REGISTRY_USERNAME --docker-password $REGISTRY_TOKEN --docker-server https://docker.cke-cs.com
+```
+
+## Installation
+
+Update `values.yaml` file with your ingress hostname and licence key. Next install chart in your kubernetes cluster:
```
Comment (czerwonkabartosz) [inline kubernetes/helm/ckeditor-cs-stack/README.md:27]: > your ingress hostname

What does it mean? I have to change this:  
 

```
    hosts:
      - host: ckeditor-cs.organization.test
```

but for local devepolemnt what I should provide?

File: kubernetes/helm/ckeditor-cs-stack/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+```sh
+helm dependency build
```
Comment (czerwonkabartosz) [inline kubernetes/helm/ckeditor-cs-stack/README.md:30]:
```
helm dependency build             
zsh: command not found: helm
```

:cry: 

IMO we could add the requirements section and add them information about minikub, helm ,etc.

Comment (czerwonkabartosz) [reply kubernetes/helm/ckeditor-cs-stack/README.md:30]:
```
helm dependency build                  
Error: no repository definition for https://charts.bitnami.com/bitnami, https://charts.bitnami.com/bitnami, https://charts.bitnami.com/bitnami. Please add the missing repos via 'helm repo add'
```

Comment (czerwonkabartosz) [reply kubernetes/helm/ckeditor-cs-stack/README.md:30]: I had to run this command: \`helm repo add bitnami https://charts.bitnami.com/bitnami`

File: kubernetes/helm/ckeditor-cs-stack/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+helm install ckeditor-cs-stack .
+```
+
+## Verify installation
```
Comment (czerwonkabartosz) [inline kubernetes/helm/ckeditor-cs-stack/README.md:34]: Container restart and restarts

Comment (czerwonkabartosz) [reply kubernetes/helm/ckeditor-cs-stack/README.md:34]:
```
{"level":50,"time":"2021-05-11T07:08:33.018Z","msg":"getaddrinfo ENOTFOUND ckeditor-cs-mysql","data":{"name":"Error","message":"getaddrinfo ENOTFOUND ckeditor-cs-mysql","errno":"ENOTFOUND","code":"ENOTFOUND","syscall":"getaddrinfo","hostname":"ckeditor-cs-mysql","fatal":true}}
```

**Review by czerwonkabartosz — CHANGES_REQUESTED** [review]

**Comment by czerwonkabartosz** [comment]
One more thing but mainly for @SlawekGreczyn IMO we could add a new section to our docs Installation / K8 



We could also think about adding Installation/AWS.

File: kubernetes/helm/ckeditor-cs/values.yaml [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  annotations: {}
+  # The name of the service account to use. If not set and create is true, a
+  # name is generated using the fullname template
+  name: ""
+
+secret:
+  # The name of the secret to use. If not set, a name is generated using the
+  # fullname template
+  name: ""
+  # Whole list of container environment variables is accessible here
```
Comment (cmwd) [inline kubernetes/helm/ckeditor-cs/values.yaml:27]: This information should be in the readme

File: kubernetes/helm/ckeditor-cs/values.yaml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  annotations: {}
+  hosts:
+    - host: ckeditor-cs.organization.test
+      paths:
+      - path: /
+  tls: []
+  #  - secretName: ckeditor-cs-organization-test hosts:
+  #      - ckeditor-cs.organization.test
+
+# Default resources request is set up for 250 concurrent users, for more
```
Comment (cmwd) [inline kubernetes/helm/ckeditor-cs/values.yaml:68]: ^^

**Review by cmwd — CHANGES_REQUESTED** [review]
I haven't tested the script itself, yet. But I think we should improve the docs. If you need any help let me know

File: kubernetes/helm/ckeditor-cs/values.yaml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+# information follow there:
+# https://ckeditor.com/docs/cs/latest/onpremises/cs-onpremises/requirements.html
+resources:
+  requests:
+    cpu: 1000m
+    memory: 512Mi
+
+autoscaling:
+  enabled: false
```
[the remark this answers is not in the document]
Comment (kbetanski) [reply kubernetes/helm/ckeditor-cs/values.yaml:77]: It's at the top of the file as `replicaCount`

File: kubernetes/helm/ckeditor-cs/templates/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "ckeditor-cs.fullname" . }}
+  labels:
+    {{- include "ckeditor-cs.labels" . | nindent 4 }}
+spec:
+  {{- if not .Values.autoscaling.enabled }}
```
[the remark this answers is not in the document]
Comment (kbetanski) [reply kubernetes/helm/ckeditor-cs/templates/deployment.yaml:8]: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#how-does-the-horizontal-pod-autoscaler-work
tl;dr
Autoscaler works as a control loop fetching container metrics (cpu and memory usage) in 15s interval. Then it calculates the percentage of resource usage from comparing values from those metrics against `resources.requests`. That's the default configuration, with some tweaking around it can also work with custom metrics.

File: kubernetes/helm/ckeditor-cs-stack/README.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
+# CKEditor Collaboration Server On-Premise stack Helm chart
+
+>:warning: **Not for production purposes!** This configuration is insecure, not persistent and its purpose is testing only.
```
[the remark this answers is not in the document]
Comment (kbetanski) [reply kubernetes/helm/ckeditor-cs-stack/README.md:3]: @cmwd Main question is: should we support running third-party services in containers and its configuration? Afaik our policy is to recommend running them in managed services in cloud e.g. AWS's Aurora

File: kubernetes/helm/ckeditor-cs-stack/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (kbetanski) [reply kubernetes/helm/ckeditor-cs-stack/README.md:27]: @czerwonkabartosz Do you mean storing license key? I can add information about making separate file that can be encrypted before storing it in git repository or elsewhere.

File: kubernetes/helm/ckeditor-cs-stack/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (czerwonkabartosz) [reply kubernetes/helm/ckeditor-cs-stack/README.md:3]: Generally, we prefer to not recommend using DBs as containers. This `ckeditor-cs-stack` in only for development purposes.   
  
We should be inspired by GitLab in this case [https://docs.gitlab.com/charts/installation/deployment.html#postgresql](https://docs.gitlab.com/charts/installation/deployment.html#postgresql)

BTW. Maybe the name is wrong? `ckeditor-cs-dev-stack` ?

File: kubernetes/helm/ckeditor-cs-stack/README.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+ENVIRONMENTS_MANAGEMENT_SECRET_KEY="$(kubectl get secret ckeditor-cs -o json | jq -r '.data.ENVIRONMENTS_MANAGEMENT_SECRET_KEY' | base64 -d)"
+
+kubectl run ckeditor-cs-tests -i --rm \
+  --image docker.cke-cs.com/cs-tests:$VERSION \
+  --env APPLICATION_ENDPOINT="$APPLICATION_ENDPOINT" \
+  --env ENVIRONMENTS_MANAGEMENT_SECRET_KEY="$ENVIRONMENTS_MANAGEMENT_SECRET_KEY" \
+  --overrides='{ "spec": { "imagePullSecrets": [{"name": "docker-cke-cs-com"}] } }' \
+  --restart='Never'
+```
+Documentation: [Testing onpremise](https://ckeditor.com/docs/cs/latest/onpremises/cs-onpremises/testing/docker.html)
```
[the remark this answers is not in the document]
Comment (czerwonkabartosz) [reply kubernetes/helm/ckeditor-cs-stack/README.md:51]: :heart:

**Review by czerwonkabartosz — CHANGES_REQUESTED** [review]
One more thing. Could you add support for creating a collaboration worker as a separate container?

https://ckeditor.com/docs/cs/latest/onpremises/cs-onpremises/worker.html

File: kubernetes/helm/ckeditor-cs-stack/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kbetanski) [reply kubernetes/helm/ckeditor-cs-stack/README.md]: @czerwonkabartosz Changed it to `ckeditor-cs-development-stack`

File: kubernetes/helm/README.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+Server. Used charts as a dependencies:
+- `ckeditor-cs`
+- `bitnami/mysql`
+- `bitnami/redis`
+
+>:warning: **Not for production purposes!** This configuration is insecure, not
+>persistent and its purpose is testing only.
+
+## Quick start
+
```
Comment (czerwonkabartosz) [inline kubernetes/helm/README.md:30]: IMO it will be good to add here information that this is a Quick Start for production and it required created databases. Or something similar.

Comment (czerwonkabartosz) [reply kubernetes/helm/README.md:30]: Or maybe you could move it to `ckeditor-cs/READM.md` and here add links only:

*   Dev Quick Start 
*   Production Quick Start 

Or something similar.

File: kubernetes/helm/ckeditor-cs/values.yaml [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+## Override for fullname template
+nameOverride: ""
+
+server:
+  ## Name of the server component container
+  name: server
+
+  ## Number of desired collaboration server instances
+  replicaCount: 1
```
Comment (czerwonkabartosz) [inline kubernetes/helm/ckeditor-cs/values.yaml:13]: IMO we could change it to 2.

File: kubernetes/helm/README.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+```
+
+- test deployment
+```sh
+kubectl run ckeditor-cs-tests -i --rm \
+    --image docker.cke-cs.com/cs-tests:"$(kubectl get deployments.apps ckeditor-cs-server -o json | jq -r '.spec.template.spec.containers[0].image' | sed 's/.*://')" \
+    --env APPLICATION_ENDPOINT="$(kubectl get ingresses.networking.k8s.io ckeditor-cs-server -o json | jq -r '.spec.rules[0].host' | sed 's|^|http://|')" \
+    --env ENVIRONMENTS_MANAGEMENT_SECRET_KEY="$(kubectl get secret ckeditor-cs-server -o json | jq -r '.data.ENVIRONMENTS_MANAGEMENT_SECRET_KEY' | base64 -d)" \
+    --overrides='{ "spec": { "imagePullSecrets": [{"name": "docker-cke-cs-com"}] } }' \
+    --restart='Never'
```
Comment (czerwonkabartosz) [inline kubernetes/helm/README.md:61]: IMO separate file for it can work quite good ;)

File: kubernetes/helm/ckeditor-cs-development-stack/README.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+helm repo update
+helm dependency update
+helm install ckeditor-cs .
+```
+
+Deleting it:
+```sh
+helm delete ckeditor-cs
+```
+## Local environment
```
Comment (czerwonkabartosz) [inline kubernetes/helm/ckeditor-cs-development-stack/README.md:39]: It looks the container is running for me, but I'm :baby:  in K8 and I don't know how I can check it manually - I don't know URL for running instance.   
  
And will be great to have a section with instruction on how you can test it. We can use approach with running tests `kubectl run ckeditor-cs-tests -i --rm \`

File: kubernetes/helm/ckeditor-cs-development-stack/README.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+./init.sh
+```
+
+Create `dev.values.yaml` file with license key:
+```yaml
+ckeditor-cs:
+  server:
+    secret:
+      data:
+        LICENSE_KEY: xxx
```
Comment (czerwonkabartosz) [inline kubernetes/helm/ckeditor-cs-development-stack/README.md:56]: WDYT about adding this file to the repo?

**Review by czerwonkabartosz — CHANGES_REQUESTED** [review]
Looks cool for me 👍  I've added a few small comments.

File: kubernetes/helm/ckeditor-cs-development-stack/README.md (hunk 14 above)
Comment (kbetanski) [reply kubernetes/helm/ckeditor-cs-development-stack/README.md:56]: I'd like to keep it in `.gitignore`. This way we can minimalize threat of commiting sensitive data to the repository.

File: kubernetes/helm/ckeditor-cs-development-stack/README.md (hunk 13 above)
Comment (kbetanski) [reply kubernetes/helm/ckeditor-cs-development-stack/README.md:39]: Added a separate file with script running e2e tests. As for URL, there is now information about default hostname from which collaboration server should be accessible by default.

File: kubernetes/helm/README.md (hunk 12 above)
Comment (kbetanski) [reply kubernetes/helm/README.md]: Moved to `/kubernetes/helm/test-deployment.sh`

File: kubernetes/helm/ckeditor-cs-development-stack/README.md (hunk 14 above)
Comment (czerwonkabartosz) [reply kubernetes/helm/ckeditor-cs-development-stack/README.md]: > I'd like to keep it in `.gitignore`. This way we can minimalize threat of commiting sensitive data to the repository.

But it is more useful/convenient for the customers IMO.

**Review by czerwonkabartosz — APPROVED** [review]
LGTM 🎉  Amazing work 😄

**Comment by kbetanski** [comment]
@LukaszGudel gave this PR a 👍. @cmwd Do you want to review this PR after all fixes, or can I merge it?
